src/openzeppelin/account.cairo

```diff
@@ -0,0 +1,234 @@
+    ) -> felt252 {
+        _validate_transaction()
+    }
+
+    #[external]
+    fn set_public_key(new_public_key: felt252) {
+        _assert_only_self();
+        public_key::write(new_public_key);
+    }
+
+    //
+    // View
+    //
+
+    #[view]
+    fn get_public_key() -> felt252 {
+        public_key::read()
+    }
+
+    // todo: fix Span serde
+    // #[view]
+    fn is_valid_signature(message: felt252, signature: Span<felt252>) -> u32 {
+        if _is_valid_signature(message, signature) {
+            ERC1271_VALIDATED
+        } else {
+            0_u32
+        }
+    }
+
+    #[view]
+    fn supports_interface(interface_id: u32) -> bool {
+        ERC165Contract::supports_interface(interface_id)
+    }
+
+    //
+    // Internals
+    //
+
+    fn _assert_only_self() {
+        let caller = get_caller_address();
+        let self = get_contract_address();
+        assert(self == caller, 'Account: unauthorized');
+    }
+
+    fn _validate_transaction() -> felt252 {
+        let tx_info = get_tx_info().unbox();
+        let tx_hash = tx_info.transaction_hash;
+        let signature = tx_info.signature;
+        assert(_is_valid_signature(tx_hash, signature), 'Account: invalid signature');
+        starknet::VALIDATED
+    }
+
+    fn _is_valid_signature(message: felt252, signature: Span<felt252>) -> bool {
+        let valid_length = signature.len() == 2_u32;
+        
+        valid_length & check_ecdsa_signature(
+            message,
+            public_key::read(),
+            *signature.at(0_u32),
+            *signature.at(1_u32)
+        )
+    }
+
+    fn _execute_calls(mut calls: Array<Call>, mut res: Array<Span<felt252>>) -> Array<Span<felt252>> {
+        check_gas();
+        match calls.pop_front() {
+            Option::Some(call) => {
+                let _res = _execute_single_call(call);
+                res.append(_res);
+                return _execute_calls(calls, res);
+            },
+            Option::None(_) => {
+                return res;
+            },
+        }
+    }
+
+    fn _execute_single_call(mut call: Call) -> Span<felt252> {
+        let Call{to, selector, calldata } = call;
+        starknet::call_contract_syscall(to, selector, calldata.span()).unwrap_syscall()
+    }
+}
+
+impl ArrayCallDrop of Drop::<Array<Call>>;
+
+impl CallSerde of Serde::<Call> {
+    fn serialize(ref output: Array<felt252>, input: Call) {
+        let Call{to, selector, calldata } = input;
+        Serde::serialize(ref output, to);
+        Serde::serialize(ref output, selector);
+        Serde::serialize(ref output, calldata);
+    }
+
+    fn deserialize(ref serialized: Span<felt252>) -> Option<Call> {
+        let to = Serde::<ContractAddress>::deserialize(ref serialized)?;
+        let selector = Serde::<felt252>::deserialize(ref serialized)?;
+        let calldata = Serde::<Array::<felt252>>::deserialize(ref serialized)?;
+        Option::Some(Call { to, selector, calldata })
+    }
+}
```

Comment:
`Serde` is now derivable on structs on the latest cairo preview

Reply:
this comes from the account branch that this PR is based on. If relevant, please raise this comment on #583